modules/mauth-common/src/main/java/com/mdsol/mauth/util/MAuthSignatureHelper.java

```diff
@@ -206,53 +206,97 @@ public static String getHexEncodedDigestedString(byte[] unencryptedData) {
-        String value = keyPair.length > 1 ? param.split("=")[1] : "";
-        if (encryptedQueryParams.length() > 0) {
-          encryptedQueryParams.append('&');
-        }
-        encryptedQueryParams.append(urlEncodeValue(name)).append('=').append(urlEncodeValue(value));
-      }
-    }
-
-    return encryptedQueryParams.toString();
+    return Arrays.stream(encodedQuery.split("&"))
+        .filter(s -> !s.isEmpty())
+        .map(keyValStr -> {
+          String[] split = keyValStr.split("=");
+          String key = split[0];
+          String value = split.length > 1 ? split[1] : "";
+          return Pair.of(urlDecodeValue(key), urlDecodeValue(value));
+        })
+        .sorted()
+        .map(keyVal -> urlEncodeValue(keyVal.getKey()) + "=" + urlEncodeValue(keyVal.getValue()))
+        .collect(Collectors.joining("&"));
   }
 
   /**
-   * encode a string value using `UTF-8` encoding scheme
+   * encode a string using `UTF-8` encoding scheme
    * @param value
-   * @return encoded String
+   * @return the encoded String
    *
-   * See https://stackoverflow.com/questions/10786042/java-url-encoding-of-query-string-parameters
+   * See https://stackoverflow.com/questions/13060034/what-is-correct-oauth-percent-encoding
    */
   private static String urlEncodeValue(String value) {
     if (value == null ||  value.isEmpty())
       return value;
 
     try {
       String encodedValue = URLEncoder.encode(value, StandardCharsets.UTF_8.toString());
-      encodedValue = encodedValue.replace("+", "%20");
-      encodedValue = encodedValue.replace("%7E", "~");
-      encodedValue = encodedValue.replace("*", "%2A");
+      // OAuth encodes some characters differently
+      encodedValue = encodedValue.replace("+", "%20").replace("%7E", "~").replace("*", "%2A");
+      return encodedValue;
+    } catch (UnsupportedEncodingException ex) {
+      throw new RuntimeException(ex.getCause());
+    }
+  }
+
+  /**
+   * decode a string using `UTF-8` scheme
+   * @param encodedValue
+   * @return the decoded String
+   *
+   * See https://docs.oracle.com/javase/8/docs/api/java/net/URLDecoder.html
+   */
+  private static String urlDecodeValue(String encodedValue) {
+    if (encodedValue == null || encodedValue.isEmpty())
       return encodedValue;
+
+    try {
+      String data = encodedValue.replaceAll("%(?![0-9a-fA-F]{2})", "%25").replaceAll("\\+", " ");
+      return URLDecoder.decode(data, StandardCharsets.UTF_8.toString());
     } catch (UnsupportedEncodingException ex) {
       throw new RuntimeException(ex.getCause());
     }
   }
 
+  /**
+   * normalize  url-encoded path string
+   * @param encodedPath
+   * @return the normalized string of path
+   */
+  public static String normalizePath(String encodedPath) {
+    if (encodedPath == null || encodedPath.isEmpty())
+      return "";
+
+    //Normalize percent encoding to uppercase i.e. %cf%80 => %CF%80
+    Pattern pattern = Pattern.compile("%[a-f0-9]{2}");
```

Comment:
The point of compiling a static pattern is so that you can reuse it across multiple calls. Defeats the purpose if you're recompiling it every call to `normalizePath` doesn't it? :)

Reply:
moved the static pattern out